primefaces/src/main/java/org/primefaces/util/PropertyDescriptorResolver.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2009-2023 PrimeTek Informatics
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.primefaces.util;
+
+import java.beans.IntrospectionException;
+import java.beans.PropertyDescriptor;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.regex.Pattern;
+import javax.faces.FacesException;
+
+public interface PropertyDescriptorResolver {
+
+    PropertyDescriptor get(Class<?> klazz, String name);
+
+    Object getValue(Object bean, String name);
+
+    void flush();
+
+    class DefaultResolver implements PropertyDescriptorResolver {
+
+        private static final Pattern NESTED_EXPRESSION_PATTERN = Pattern.compile("\\.");
+        private final Map<String, Map<String, PropertyDescriptor>> pdCache;
+
+        public DefaultResolver() {
+            pdCache = new ConcurrentHashMap<>();
+        }
+
+        @Override
+        public PropertyDescriptor get(Class<?> klazz, String expression) {
+            PropertyDescriptor pd = null;
+            Class<?> parent = klazz;
+
+            for (String field : NESTED_EXPRESSION_PATTERN.split(expression)) {
```

Comment:
this can lead to problems
if you check the current ReflectionLazyDataModel#getFieldValue, it loops on the object instances instead of of the property-descriptor

if you have e.g.
company.user.field1

class Company {
    private User user;
}

but the only the actual user instance, (e.g. a class SpecializedUser) has field1, the current code will fail?
we need some tests here

Reply:
`ReflectionLazyDataModel#getFieldValue `does the same thing as `DefaultResolver#getValue`. You actually didn't need to call `getFieldType` since you got the value before. Now about the issues you mentionned, they're present for `JPALazyDataModel` (whether you use the new resolver or the present master code)

If inheritance present, it gets tricky to determine property types with only `entityClass`

Reply:
JPALazyDataModel is something different, you dont filter/sort on existing objects, you build JPA expressions. you need to rely on class infos.

in case of the reflection, you can resolve/filter/sort on company.user.field1 if you work on the current object tree
we should just make sure via tests that it works, therfore it wa more like a question :D

Reply:
> in case of the reflection, you can resolve/filter/sort on company.user.field1 if you work on the current object tree
we should just make sure via tests that it works, therfore it wa more like a question :D

I know

> JPALazyDataModel is something different, you dont filter/sort on existing objects, you build JPA expressions. you need to rely on class infos.

Yup, but as I am not a JPA user (at least not using JPQL, and criteria stuff) i'd rather let you know. If you take that sample

https://github.com/primefaces/primefaces/blob/4ad098c54c8d22f482e4e6ed3be3945ed32c230e/primefaces/src/main/java/org/primefaces/model/JpaLazyDataModel.java#L384-L396

If `A` is abstract (no attributes) and `B` extends `A` and have an `id`, and you configure `A` as your `entityClass` and `id` as your rowKey, you'll have a problem. Same thing with field:

https://github.com/primefaces/primefaces/blob/4ad098c54c8d22f482e4e6ed3be3945ed32c230e/primefaces/src/main/java/org/primefaces/model/JpaLazyDataModel.java#L175

I don't think there is an easy way to workaround this, maybe worth to mention as a limitation in JPALazyDataModel